### DISCUSSION TITLE
auth: ldap Where/what Cert with ldaps to an AD

### DISCUSSION THREAD

**Discussion by simpsonetti** [Ylianst/MeshCentral#5758] [Q&A]
Hi, 
i want to auth my users against an Active Directory. So for this i setup the auth : ldap part and add as url ldaps://IP:636 . But this didn't work. 
So i start with --debug ldap and i get this here:
LDAP: LDAP Error: Error: unable to get local issuer certificate
LDAP: LDAP OnError: Error: unable to get local issuer certificate
LDAP: LDAP OnError: Error: unable to get local issuer certificate

So is there an option where i can configure that and which cert did MesC expect here?

**Comment by si458** [comment]
you are trying to connect to your AD over secure, try using `ldap://IP:389`
i cant remember if mesh supports ldaps or not? just looking into config now

**Reply by si458** [reply]
also the is a video here for LDAP setup - https://www.youtube.com/watch?v=UQMbNQk85g8

**Reply by simpsonetti** [reply]
i'm watching this video and configured. but i don't know, if our AD is accepting LDAP at port 389. i don't think so. 
So could i setup to disable the check for ssl?

**Reply by si458** [reply]
I'm not sure I'm awfk at mo, will look later

**Reply by simpsonetti** [reply]
okay, that would be nice. 
i think with "tlsOptions": { "rejectUnauthorized": false }, this will disable check?!
So far i don't get the erro message, now my user woudn't be found. 

But, i don't read this anywhere in docs, videos etc. : when enabling ldap, the first user that was created (the main administrator) can't login? How do i configure MC then?

**Reply by si458** [reply]
we use `ldapauth-fork` for the ldap auth
`ldapauth-fork` says to use `tlsOptions` to set any TLS requirements - https://nodejs.org/docs/latest-v10.x/api/tls.html#tls_tls_connect_options_callback
so i believe you would do 
```
"ldapOptions" : {
    "tlsOptions": {
        "rejectUnauthorized": false
    }
}
```

**Reply by si458** [reply]
you might also need to share your config.json as setting up LDAP can be confusing...

EDIT:
this was my sample one i used with openldap 
```
{
    "LDAPOptions": {
        "url": "ldap://localhost:389",
        "bindDN": "cn=admin,dc=example,dc=org",
        "bindCredentials": "adminpassword",
        "searchBase": "ou=StaffUsers,dc=example,dc=org",
        "searchFilter": "(uid={{username}})",
        "reconnect": true,
        "tlsOptions": {
            "rejectUnauthorized": false
        }
    },
    "ldapUserKey": "uid",
    "ldapUserName": "uid",
    "ldapUserRealName": "displayName",
    "_ldapSiteAdminGroups": "500",
    "_ldapUserGroups": "gidNumber",
    "_ldapSyncWithUserGroups": true,
    "ldapSaveUserToFile": "../meshcentral-data/ldap.txt"
}
```

**Reply by simpsonetti** [reply]
here is my domainpart config with the ldap:
I'm just starting this server as a productive version and i'm just configuring the config.json. So until now, there are not all options configured i need. Step by step...
```
{
    "auth": "ldap",
    "ldapOptions": {
        "url": "ldaps://172.16.20.1:636/",
        "tlsOptions": {
            "rejectUnauthorized": false
        },
        "bindDN": "CN=sk-meshcentral-ad,OU=MeshCentral,OU=Servicekonten,OU=BS,OU=infra,DC=domain,DC=com",
        "bindCredentials": "securepass",
        "searchBase": "DC=domain,DC=com",
        "searchFilter": "(name={{cn}})"
    },
    "ldapUserKey": "cn",
    "_ldapUserBinaryKey": "objectSid",
    "ldapUserName": "{{{givenName}}} {{{name}}}",
    "ldapUserEmail": "mail",
    "ldapUserRealName": "{{{givenName}}} {{{name}}}",
    "ldapUserPhoneNumber": "telephoneNumber",
    "ldapUserImage": "thumbnailPhoto",
    "ldapSaveUserToFile": "../meshcentral-data/ldapusers.txt",
    "meshQuota": 100000000,
    "userQuota": 1000000
}
```

**Reply by simpsonetti** [reply]
Hi @si458 , do you check MC against a regular Active Directory oder only openLDAP?

**Reply by si458** [reply]
i dont use ldap anymore, but the config above was just my testing config was i was looking into a ldap bug, but i setup and used openldap and not AD

**Reply by simpsonetti** [reply]
so, after hours of investigation:
to ask our AD i need to setup CAcert. That's my cause here. "Regular" ldapsearch" doesn't work too. So i give him the "LDAPTLS_CACERT" env and than my ldapsearch works. 

So for meshcentral, there has to be an option, so specify my CACERT. But it seems, there isn't anything for this?
@si458 : i edit my config.json via VS Code, so the autocompletion is working with the "$schema" and therefor is no option!? 
Could you check this?

**Reply by si458** [reply]
sadly the is no option for certificates for LDAP after quickly checking the code,
all you can set is `ldapOptions: { }` and inside that, you put extra options to pass to `ldapauth-fork`
so the best thing i can suggest is to ask in the [module repo](https://github.com/vesse/node-ldapauth-fork) and see if they can say HOW you should add a certificate in to do authentication with CA certificates.
once you have it working there normally in nodejs, then we can add it into meshcentral
